ZMP generalization

VHchavez opened the pull request:
## Description
Improves on ZMP Method:
- Lambda parameter is now a list that can be used to extrapolate the potential iterative. 
- Adds mixing parameter to regulate how much of the potential get added with each new lambda. Default 1 works fine. 
- Adds generalization of functionals -> Exponential and Logarithmic.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] Iterative calculation only works on Hartree functional. Other functionals do not converge reliably. But for a single lambda parameter, they work just fine. 

## Status
- [ ] Ready to go

ymshi449 commented on n2v/methods/zmp.py:
```diff
@@ -33,174 +35,269 @@ def zmp(self,
-                    diis_error = 2 * diis_error_a
+                Fa = np.zeros((self.nbf, self.nbf))
+                Fb = np.zeros((self.nbf, self.nbf))
+                J, _ = self.form_jk(Cocca, Coccb)
```
This is not necessary for zmp_functional.lower() == 'hartree' right?

VHchavez replied:
True

ymshi449 commented on n2v/methods/zmp.py:
```diff
@@ -33,174 +35,269 @@ def zmp(self,
-                        state_vectors_b.append(Fb.copy())
-                        error_vectors_b.append(grad_b.copy())
+                    #Equations 37, 38, 39, 40 of Reference (3)
+                    vc = self.generate_s_fucntional(lam_i, zmp_functional)
```
Would it be better to also include condition zmp_functional.lower() == 'hartree' in generate_s_fucntional() instead of picking this out?

VHchavez replied:
Makes sense. Will put inside

ymshi449 commented on n2v/methods/zmp.py:
```diff
@@ -33,174 +35,269 @@ def zmp(self,
+            Cocca = self.ct[0]
+            Coccb = self.ct[1]
+            D_old = self.nt[0]
+            vc_old = np.zeros_like(Da)
```
Probably we don't need to zero vc for each lambda. Instead, vc of old lambda should contribute to vc of current lambda.

ymshi449 commented on n2v/methods/zmp.py:
```diff
@@ -33,174 +35,269 @@ def zmp(self,
+                vc = (1-self.mixing) * vc + (self.mixing) * vc_old
+
+                #Equation 10 of Reference (1)
+                Fa += self.T + self.V + self.va + vc + vc_global
```
I am a little confused. What is vc_global?

ymshi449 replied:
I see. It is vc of last lambda.

VHchavez replied:
Yup, it is just wording. Wanted to keep current_vc from old_vc to format as in the paper.

VHchavez commented:
Yeah, @ymshi449, the iterations are working. 
I have been using equation 28 since this is what they have on all the other ZMP papers. I am uncertain about equation (43) to be honest. 

The only thing missing in the code is that the non-Hartree methods force the -1/r behaviour for the potential generated. 
I haven't done this since I find it quite ugly... But not sure what to do. Should I add an (error function) * (-1/r ) for large r to the other functionals?

VHchavez commented:
Alright, examples should be working. 
Extra functionals are there, and they produce a solution with a single lambda. Iterative lambda won't get very far. 
Made a few minor changes and fixed the LGTM warnings. 
Fixed the badges for the main README, to remind us that none of the code is tested :) Will work on that as soon as this is merged.
